src/components/Table/RowState.ts

```diff
@@ -37,6 +35,82 @@ export class RowState {
     makeAutoObservable(this);
   }
 
+  get selectedIds(): string[] {
+    // Return only ids that are fully checked, i.e. not partial
+    const ids = [...this.selectedRows.entries()].filter(([, v]) => v === "checked").map(([k]) => k);
+    // Hide our header marker
+    const headerIndex = ids.indexOf("header");
+    if (headerIndex > -1) {
+      ids.splice(headerIndex, 1);
+    }
+    return ids;
+  }
+
+  // Should be called in an Observer/useComputed to trigger re-renders
+  getSelected(id: string): SelectedState {
+    // We may not have every row in here, i.e. on 1st page load or after clicking here, so assume unchecked
+    return this.selectedRows.get(id) || "unchecked";
+  }
+
+  selectRow(id: string, selected: boolean): void {
+    if (id === "header") {
+      // Select/unselect all has special behavior
+      if (selected) {
+        // Just mash the header + all rows + children as selected
+        const map = new Map<string, SelectedState>();
+        map.set("header", "checked");
+        const todo = [...this.rows.current];
+        while (todo.length > 0) {
+          const r = todo.pop()!;
+          map.set(r.id, "checked");
+          if (r.children) {
+            todo.push(...r.children);
+          }
+        }
+        this.selectedRows.replace(map);
+      } else {
+        // Similarly "unmash" all rows + children.
+        this.selectedRows.clear();
+      }
+    } else {
+      // This is the regular/non-header behavior to just add/remove the individual row id,
+      // plus percolate the change down-to-child + up-to-parents.
+
+      // Find the clicked on row
+      const curr = findRow(this.rows.current, id);
+      if (!curr) {
+        return;
+      }
+
+      // Everything here & down is deterministically on/off
+      const map = new Map<string, SelectedState>();
+      const todo = [curr.row];
+      while (todo.length > 0) {
+        const row = todo.pop()!;
+        map.set(row.id, selected ? "checked" : "unchecked");
+        if (row.children) {
+          todo.push(...row.children);
+        }
+      }
```

Comment:
Dang. If only we could tell Array.prototype.flat() what key to flatten on this could be so much more succinct. 

```js
const map = new Map<string, SelectedState>();
curr.row.flatIshMap((row) => map.set(row.id, selected ? "checked" : "unchecked"));
```

Not generally a fan of while/for loops as they're more imperative than declarative. But, they work, so can't really fuss too much.

I do feel like a helper function could assist?

```js
function traverseWithChildren(items, fn, parent) {
  items.forEach((item) => {
    fn(item);
    if (item.children) traverseWithChildren(item.children, fn, item);
  });
}

traverseWithChildren(curr.row,
  (item) => map.set(row.id, selected ? "checked" : "unchecked"),
);
```

Reply:
Ah nice, good call!

When writing the 2nd `while`, I thought of that, but swear that it was "different from the other `while`" at first; but turns out it either never was, or at least is not anymore.

So I added a `visit` method giving the "visitor pattern" is exactly what this sort of "please traverse this complex (okay, for us just a tree) data structure for me, and just call my function as necessary kthx" pattern is typically called:

```
function visit(rows: GridDataRow<any>[], fn: (row: GridDataRow<any>) => void): void {
  const todo = [...rows];
  while (todo.length > 0) {
    const row = todo.pop()!;
    fn(row);
    if (row.children) {
      todo.push(...row.children);
    }
  }
}
```

Pushed up.